src/cryptsetup/cryptsetup.c

```diff
@@ -2265,6 +2269,44 @@ static void remove_and_erasep(const char **p) {
                 log_warning_errno(r, "Unable to erase key file '%s', ignoring: %m", *p);
 }
 
+static const char* determine_token_type(void) {
+        if (arg_tpm2_device || arg_tpm2_device_auto)
+                return "tpm2";
+        else if (arg_fido2_device || arg_fido2_device_auto)
+                return "fido2";
+        else if (arg_pkcs11_uri || arg_pkcs11_uri_auto)
+                return "pkcs11";
+        return NULL;
+}
+
+static int discover_key(const char *key_file, const char *volume, const char *token_type, struct iovec *ret_key_data) {
+        _cleanup_free_ char *bindname = NULL;
+        int r;
+
```

Comment:
we usually add `assert()`s to the top of cuntions validating parameters to be non-NULL that are mandatory. In particular return parameters we unconditionally initialize should be reated that way.